In [42]:
import pandas as pd

# Read the three files into python as dataframes

In [43]:
file1 = pd.read_csv (r"C:\Users\fabi_\OneDrive\Estudos e Cursos\Data analytics\Ironhack.Labs\lab-customer-analysis-round-1\files_for_lab\csv_files\file1.csv")

file2 = pd.read_csv (r"C:\Users\fabi_\OneDrive\Estudos e Cursos\Data analytics\Ironhack.Labs\lab-customer-analysis-round-1\files_for_lab\csv_files\file2.csv")

file3 = pd.read_csv (r"C:\Users\fabi_\OneDrive\Estudos e Cursos\Data analytics\Ironhack.Labs\lab-customer-analysis-round-1\files_for_lab\csv_files\file3.csv")

# Show the DataFrame's shape

In [44]:
file1

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [45]:
file1.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

In [46]:
file2

Customer          ST GENDER Education Customer Lifetime Value  Income  \
0    GS98873     Arizona      F  Bachelor              323912.47%   16061   
1    CW49887  California      F    Master              462680.11%   79487   
2    MY31220  California      F   College              899704.02%   54230   
3    UH35128      Oregon      F   College             2580706.30%   71210   
4    WH52799     Arizona      F   College              380812.21%   94903   
..       ...         ...    ...       ...                     ...     ...   
991  HV85198     Arizona      M    Master              847141.75%   63513   
992  BS91566     Arizona      F   College              543121.91%   58161   
993  IL40123      Nevada      F   College              568964.41%   83640   
994  MY32149  California      F    Master              368672.38%       0   
995  SA91515  California      M  Bachelor              399258.39%       0   

     Monthly Premium Auto Number of Open Complaints  Total Claim Amount  \
0                      88                    1/0/00          633.600000   
1                     114                    1/0/00          547.200000   
2                     112                    1/0/00          537.600000   
3                     214                    1/1/00         1027.200000   
4                      94                    1/0/00          451.200000   
..                    ...                       ...                 ...   
991                    70                    1/0/00          185.667213   
992                    68                    1/0/00          140.747286   
993                    70                    1/0/00          471.050488   
994                    96                    1/0/00           28.460568   
995                   111                    1/0/00          700.349052   

        Policy Type  Vehicle Class  
0     Personal Auto  Four-Door Car  
1      Special Auto            SUV  
2     Personal Auto   Two-Door Car  
3     Personal Auto     Luxury Car  
4    Corporate Auto   Two-Door Car  
..              ...            ...  
991   Personal Auto  Four-Door Car  
992  Corporate Auto  Four-Door Car  
993  Corporate Auto   Two-Door Car  
994   Personal Auto   Two-Door Car  
995   Personal Auto            SUV  

[996 rows x 11 columns]

In [47]:
file2.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                         int64
Monthly Premium Auto           int64
Number of Open Complaints     object
Total Claim Amount           float64
Policy Type                   object
Vehicle Class                 object
dtype: object

In [48]:
file3

Customer       State  Customer Lifetime Value             Education  \
0     SA25987  Washington              3479.137523  High School or Below   
1     TB86706     Arizona              2502.637401                Master   
2     ZL73902      Nevada              3265.156348              Bachelor   
3     KX23516  California              4455.843406  High School or Below   
4     FN77294  California              7704.958480  High School or Below   
...       ...         ...                      ...                   ...   
7065  LA72316  California             23405.987980              Bachelor   
7066  PK87824  California              3096.511217               College   
7067  TD14365  California              8163.890428              Bachelor   
7068  UP19263  California              7524.442436               College   
7069  Y167826  California              2611.836866               College   

     Gender  Income  Monthly Premium Auto  Number of Open Complaints  \
0         M       0                   104                          0   
1         M       0                    66                          0   
2         F   25820                    82                          0   
3         F       0                   121                          0   
4         M   30366                   101                          2   
...     ...     ...                   ...                        ...   
7065      M   71941                    73                          0   
7066      F   21604                    79                          0   
7067      M       0                    85                          3   
7068      M   21941                    96                          0   
7069      M       0                    77                          0   

         Policy Type  Total Claim Amount  Vehicle Class  
0      Personal Auto          499.200000   Two-Door Car  
1      Personal Auto            3.468912   Two-Door Car  
2      Personal Auto          393.600000  Four-Door Car  
3      Personal Auto          699.615192            SUV  
4      Personal Auto          484.800000            SUV  
...              ...                 ...            ...  
7065   Personal Auto          198.234764  Four-Door Car  
7066  Corporate Auto          379.200000  Four-Door Car  
7067  Corporate Auto          790.784983  Four-Door Car  
7068   Personal Auto          691.200000  Four-Door Car  
7069  Corporate Auto          369.600000   Two-Door Car  

[7070 rows x 11 columns]

In [49]:
file3.dtypes

Customer                      object
State                         object
Customer Lifetime Value      float64
Education                     object
Gender                        object
Income                         int64
Monthly Premium Auto           int64
Number of Open Complaints      int64
Policy Type                   object
Total Claim Amount           float64
Vehicle Class                 object
dtype: object

# Standardize header names

In [50]:
file1.columns
cols = []
for i in range (len(file1.columns)): 
    cols.append(file1.columns[i].lower().replace(' ', '_'))
file1.columns = cols
file1 = file1.rename (columns = {'st' : 'state'})
file1

customer       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     customer_lifetime_value   income  monthly_premium_auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     number_of_open_complaints     policy_type  vehicle_class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [51]:
file2.columns
cols = []
for i in range (len(file2.columns)): 
    cols.append(file2.columns[i].lower().replace(' ', '_'))
file2.columns = cols
file2 = file2.rename (columns = {'st' : 'state'})
file2

customer       state gender education customer_lifetime_value  income  \
0    GS98873     Arizona      F  Bachelor              323912.47%   16061   
1    CW49887  California      F    Master              462680.11%   79487   
2    MY31220  California      F   College              899704.02%   54230   
3    UH35128      Oregon      F   College             2580706.30%   71210   
4    WH52799     Arizona      F   College              380812.21%   94903   
..       ...         ...    ...       ...                     ...     ...   
991  HV85198     Arizona      M    Master              847141.75%   63513   
992  BS91566     Arizona      F   College              543121.91%   58161   
993  IL40123      Nevada      F   College              568964.41%   83640   
994  MY32149  California      F    Master              368672.38%       0   
995  SA91515  California      M  Bachelor              399258.39%       0   

     monthly_premium_auto number_of_open_complaints  total_claim_amount  \
0                      88                    1/0/00          633.600000   
1                     114                    1/0/00          547.200000   
2                     112                    1/0/00          537.600000   
3                     214                    1/1/00         1027.200000   
4                      94                    1/0/00          451.200000   
..                    ...                       ...                 ...   
991                    70                    1/0/00          185.667213   
992                    68                    1/0/00          140.747286   
993                    70                    1/0/00          471.050488   
994                    96                    1/0/00           28.460568   
995                   111                    1/0/00          700.349052   

        policy_type  vehicle_class  
0     Personal Auto  Four-Door Car  
1      Special Auto            SUV  
2     Personal Auto   Two-Door Car  
3     Personal Auto     Luxury Car  
4    Corporate Auto   Two-Door Car  
..              ...            ...  
991   Personal Auto  Four-Door Car  
992  Corporate Auto  Four-Door Car  
993  Corporate Auto   Two-Door Car  
994   Personal Auto   Two-Door Car  
995   Personal Auto            SUV  

[996 rows x 11 columns]

In [52]:
file3.columns
cols = []
for i in range (len(file3.columns)): 
    cols.append(file3.columns[i].lower().replace(' ', '_'))
file3.columns = cols
file3

customer       state  customer_lifetime_value             education  \
0     SA25987  Washington              3479.137523  High School or Below   
1     TB86706     Arizona              2502.637401                Master   
2     ZL73902      Nevada              3265.156348              Bachelor   
3     KX23516  California              4455.843406  High School or Below   
4     FN77294  California              7704.958480  High School or Below   
...       ...         ...                      ...                   ...   
7065  LA72316  California             23405.987980              Bachelor   
7066  PK87824  California              3096.511217               College   
7067  TD14365  California              8163.890428              Bachelor   
7068  UP19263  California              7524.442436               College   
7069  Y167826  California              2611.836866               College   

     gender  income  monthly_premium_auto  number_of_open_complaints  \
0         M       0                   104                          0   
1         M       0                    66                          0   
2         F   25820                    82                          0   
3         F       0                   121                          0   
4         M   30366                   101                          2   
...     ...     ...                   ...                        ...   
7065      M   71941                    73                          0   
7066      F   21604                    79                          0   
7067      M       0                    85                          3   
7068      M   21941                    96                          0   
7069      M       0                    77                          0   

         policy_type  total_claim_amount  vehicle_class  
0      Personal Auto          499.200000   Two-Door Car  
1      Personal Auto            3.468912   Two-Door Car  
2      Personal Auto          393.600000  Four-Door Car  
3      Personal Auto          699.615192            SUV  
4      Personal Auto          484.800000            SUV  
...              ...                 ...            ...  
7065   Personal Auto          198.234764  Four-Door Car  
7066  Corporate Auto          379.200000  Four-Door Car  
7067  Corporate Auto          790.784983  Four-Door Car  
7068   Personal Auto          691.200000  Four-Door Car  
7069  Corporate Auto          369.600000   Two-Door Car  

[7070 rows x 11 columns]

# Rearrange the columns in the dataframe

In [53]:
# move the 'gender' column in file1 to the second position in the DataFrame
df = pd.DataFrame(file1)
gender_col = df.pop('gender')
df.insert(1, 'gender', gender_col)

file1 = df
file1

customer gender       state             education  \
0     RB50392    NaN  Washington                Master   
1     QZ44356      F     Arizona              Bachelor   
2     AI49188      F      Nevada              Bachelor   
3     WW63253      M  California              Bachelor   
4     GA49547      M  Washington  High School or Below   
...       ...    ...         ...                   ...   
4003      NaN    NaN         NaN                   NaN   
4004      NaN    NaN         NaN                   NaN   
4005      NaN    NaN         NaN                   NaN   
4006      NaN    NaN         NaN                   NaN   
4007      NaN    NaN         NaN                   NaN   

     customer_lifetime_value   income  monthly_premium_auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     number_of_open_complaints     policy_type  vehicle_class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [54]:
# organize colums in file2 so they are the same as in file1

file2 = file2[['customer','gender','state','education','customer_lifetime_value','income','monthly_premium_auto','number_of_open_complaints','policy_type','vehicle_class','total_claim_amount']]
file2

customer gender       state education customer_lifetime_value  income  \
0    GS98873      F     Arizona  Bachelor              323912.47%   16061   
1    CW49887      F  California    Master              462680.11%   79487   
2    MY31220      F  California   College              899704.02%   54230   
3    UH35128      F      Oregon   College             2580706.30%   71210   
4    WH52799      F     Arizona   College              380812.21%   94903   
..       ...    ...         ...       ...                     ...     ...   
991  HV85198      M     Arizona    Master              847141.75%   63513   
992  BS91566      F     Arizona   College              543121.91%   58161   
993  IL40123      F      Nevada   College              568964.41%   83640   
994  MY32149      F  California    Master              368672.38%       0   
995  SA91515      M  California  Bachelor              399258.39%       0   

     monthly_premium_auto number_of_open_complaints     policy_type  \
0                      88                    1/0/00   Personal Auto   
1                     114                    1/0/00    Special Auto   
2                     112                    1/0/00   Personal Auto   
3                     214                    1/1/00   Personal Auto   
4                      94                    1/0/00  Corporate Auto   
..                    ...                       ...             ...   
991                    70                    1/0/00   Personal Auto   
992                    68                    1/0/00  Corporate Auto   
993                    70                    1/0/00  Corporate Auto   
994                    96                    1/0/00   Personal Auto   
995                   111                    1/0/00   Personal Auto   

     vehicle_class  total_claim_amount  
0    Four-Door Car          633.600000  
1              SUV          547.200000  
2     Two-Door Car          537.600000  
3       Luxury Car         1027.200000  
4     Two-Door Car          451.200000  
..             ...                 ...  
991  Four-Door Car          185.667213  
992  Four-Door Car          140.747286  
993   Two-Door Car          471.050488  
994   Two-Door Car           28.460568  
995            SUV          700.349052  

[996 rows x 11 columns]

In [55]:
# organize colums in file3 so they are the same as in file1 and 2

file3 = file3[['customer','gender','state','education','customer_lifetime_value','income','monthly_premium_auto','number_of_open_complaints','policy_type','vehicle_class','total_claim_amount']]
file3

customer gender       state             education  \
0     SA25987      M  Washington  High School or Below   
1     TB86706      M     Arizona                Master   
2     ZL73902      F      Nevada              Bachelor   
3     KX23516      F  California  High School or Below   
4     FN77294      M  California  High School or Below   
...       ...    ...         ...                   ...   
7065  LA72316      M  California              Bachelor   
7066  PK87824      F  California               College   
7067  TD14365      M  California              Bachelor   
7068  UP19263      M  California               College   
7069  Y167826      M  California               College   

      customer_lifetime_value  income  monthly_premium_auto  \
0                 3479.137523       0                   104   
1                 2502.637401       0                    66   
2                 3265.156348   25820                    82   
3                 4455.843406       0                   121   
4                 7704.958480   30366                   101   
...                       ...     ...                   ...   
7065             23405.987980   71941                    73   
7066              3096.511217   21604                    79   
7067              8163.890428       0                    85   
7068              7524.442436   21941                    96   
7069              2611.836866       0                    77   

      number_of_open_complaints     policy_type  vehicle_class  \
0                             0   Personal Auto   Two-Door Car   
1                             0   Personal Auto   Two-Door Car   
2                             0   Personal Auto  Four-Door Car   
3                             0   Personal Auto            SUV   
4                             2   Personal Auto            SUV   
...                         ...             ...            ...   
7065                          0   Personal Auto  Four-Door Car   
7066                          0  Corporate Auto  Four-Door Car   
7067                          3  Corporate Auto  Four-Door Car   
7068                          0   Personal Auto  Four-Door Car   
7069                          0  Corporate Auto   Two-Door Car   

      total_claim_amount  
0             499.200000  
1               3.468912  
2             393.600000  
3             699.615192  
4             484.800000  
...                  ...  
7065          198.234764  
7066          379.200000  
7067          790.784983  
7068          691.200000  
7069          369.600000  

[7070 rows x 11 columns]

# Concatenate the three dataframes

In [56]:
merged_dfs = pd.concat ([file1, file2, file3], axis = 0)
merged_dfs

customer gender       state             education  \
0     RB50392    NaN  Washington                Master   
1     QZ44356      F     Arizona              Bachelor   
2     AI49188      F      Nevada              Bachelor   
3     WW63253      M  California              Bachelor   
4     GA49547      M  Washington  High School or Below   
...       ...    ...         ...                   ...   
7065  LA72316      M  California              Bachelor   
7066  PK87824      F  California               College   
7067  TD14365      M  California              Bachelor   
7068  UP19263      M  California               College   
7069  Y167826      M  California               College   

     customer_lifetime_value   income  monthly_premium_auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
7065             23405.98798  71941.0                  73.0   
7066             3096.511217  21604.0                  79.0   
7067             8163.890428      0.0                  85.0   
7068             7524.442436  21941.0                  96.0   
7069             2611.836866      0.0                  77.0   

     number_of_open_complaints     policy_type  vehicle_class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
7065                         0   Personal Auto  Four-Door Car   
7066                         0  Corporate Auto  Four-Door Car   
7067                         3  Corporate Auto  Four-Door Car   
7068                         0   Personal Auto  Four-Door Car   
7069                         0  Corporate Auto   Two-Door Car   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
7065          198.234764  
7066          379.200000  
7067          790.784983  
7068          691.200000  
7069          369.600000  

[12074 rows x 11 columns]

In [57]:
merged_dfs.shape

(12074, 11)

In [58]:
merged_dfs.dtypes

customer                      object
gender                        object
state                         object
education                     object
customer_lifetime_value       object
income                       float64
monthly_premium_auto         float64
number_of_open_complaints     object
policy_type                   object
vehicle_class                 object
total_claim_amount           float64
dtype: object

# Numerical and Categorigal columns

In [59]:
categ_cols = df.select_dtypes(exclude=['number']).columns.tolist()
categ_cols

['customer',
 'gender',
 'state',
 'education',
 'customer_lifetime_value',
 'number_of_open_complaints',
 'policy_type',
 'vehicle_class']

In [60]:
num_cols = df.select_dtypes(include=['number']).columns.tolist()
num_cols

['income', 'monthly_premium_auto', 'total_claim_amount']

# Understand the meaning of all columns

In [61]:
merged_dfs.nunique()

customer                     9056
gender                          5
state                           8
education                       6
customer_lifetime_value      8211
income                       5655
monthly_premium_auto          209
number_of_open_complaints      12
policy_type                     3
vehicle_class                   6
total_claim_amount           5070
dtype: int64

Regarding the columns of numerical data:
income, monthly_premium_auto, and total_claim_amount have a large number of unique values, indicating a wide range of values.

In [62]:
types_of_unique_values = ['income', 'monthly_premium_auto', 'total_claim_amount']
for column in types_of_unique_values:
    unique_values = df[column].unique()
    print(f"Unique values for {column}:")
    print(unique_values)
    print()

Unique values for income:
[    0. 48767. 36357. 62902. 55350. 14072. 28812. 77026. 99845. 83689.
 24599. 25049. 28855. 51148. 66140. 57749. 61040. 17870. 97541. 29723.
 86584. 75690. 23158. 65999. 37260. 68987. 42305. 65706. 53243. 50071.
 60021. 43244. 46896. 10105. 23218. 24804. 33190. 36014. 67163. 16701.
 46623. 64749. 16969. 11621. 17622. 11489. 66525. 33663. 22547. 61486.
 29879. 93011. 65186. 26840. 75644. 38984. 71811. 20961. 41275. 61321.
 51066. 34378. 43072. 25222. 77552. 23091. 48269. 32720. 20396. 21513.
 80744. 52822. 69738. 17780. 18107. 57740. 32627. 65795. 60475. 41837.
 32303. 40531. 79898. 56398. 71600. 27804. 45506. 99428. 20993. 37839.
 92711. 74665. 61108. 46833. 88768. 28262. 33555. 42780. 58850. 21880.
 25251. 25317. 24188. 41611. 28406. 69833. 80536. 88891. 97732. 61222.
 50335. 82210. 64495. 28859. 77330. 33060. 42557. 26372. 17514. 89270.
 29757. 51814. 24028. 28142. 52705. 54040. 22492. 21876. 70699. 27501.
 15897. 25141. 28358. 62530. 90972. 63110. 29549. 3

Regarding the columns of categorical data with a few unique values:
There may be 5 definitions for gender. Indicating possible typos or lack of standardization.
Customers may be from 8 different states.
There may be 6 educational categories.
Number of open complaints has 12 unique values, which might indicate the number of different types of complaints.
There may be 3 policy types.
There may be 6 vehicle classes.

Regarding the columns of categorical data with a big number of unique values:
Customer column has 9056 unique values out of 12074 rows, indicating possible repeated customers.
Customer lifetime value have a large number of unique values, indicating a wide range of values that probably should be converted to numerical values.

In [63]:
types_of_unique_values = ['gender','state','education','number_of_open_complaints','policy_type','vehicle_class']
for column in types_of_unique_values:
    unique_values = df[column].unique()
    print(f"Unique values for {column}:")
    print(unique_values)
    print()

Unique values for gender:
[nan 'F' 'M' 'Femal' 'Male' 'female']

Unique values for state:
['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]

Unique values for education:
['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]

Unique values for number_of_open_complaints:
['1/0/00' '1/2/00' '1/1/00' '1/3/00' '1/5/00' '1/4/00' nan]

Unique values for policy_type:
['Personal Auto' 'Corporate Auto' 'Special Auto' nan]

Unique values for vehicle_class:
['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan]



In [64]:
types_of_unique_values = ['customer','customer_lifetime_value']
for column in types_of_unique_values:
    unique_values = df[column].unique()
    print(f"Unique values for {column}:")
    print(unique_values)
    print()

Unique values for customer:
['RB50392' 'QZ44356' 'AI49188' ... 'CW49887' 'MY31220' nan]

Unique values for customer_lifetime_value:
[nan '697953.59%' '1288743.17%' ... '2031499.76%' '323912.47%'
 '899704.02%']



By looking at the actual unique categorical values we can conclude that:

For 'gender': There are typos and lack of standardization. There are also NaN values representing undefined or missing values.
For 'state': There are repeated states and lack of standardization. There are also NaN values representing undefined or missing values.
For 'education': There is lack of standard for the bachelor's degree. There are also NaN values representing undefined or missing values.
For 'number of open complaints': there seems to be inconsistent data entry or classification.There are also NaN values representing undefined or missing values.
For 'policy type': There are NaN values representing undefined or missing values.
For 'vehicle class': There are also NaN values representing undefined or missing values.
For 'customer': Column likely has repeated customers. There are also NaN values representing undefined or missing values.
For 'customer lifetime value': Probably should be converted to numerical values and standardize since some of the values are percentages. There are also NaN values representing undefined or missing values.

In [71]:
merged_dfs.isna()

customer  gender  state  customer_lifetime_value  income  \
0        False    True  False                     True   False   
1        False   False  False                    False   False   
2        False   False  False                    False   False   
3        False   False  False                    False   False   
4        False   False  False                    False   False   
...        ...     ...    ...                      ...     ...   
7065     False   False  False                    False   False   
7066     False   False  False                    False   False   
7067     False   False  False                    False   False   
7068     False   False  False                    False   False   
7069     False   False  False                    False   False   

      monthly_premium_auto  policy_type  vehicle_class  total_claim_amount  
0                    False        False          False               False  
1                    False        False          False               False  
2                    False        False          False               False  
3                    False        False          False               False  
4                    False        False          False               False  
...                    ...          ...            ...                 ...  
7065                 False        False          False               False  
7066                 False        False          False               False  
7067                 False        False          False               False  
7068                 False        False          False               False  
7069                 False        False          False               False  

[12074 rows x 9 columns]

# Data Transformation

In [65]:
# Delete the column education and the number of open complaints from the dataframe.
merged_dfs = merged_dfs.drop(['education'], axis = 1)

In [66]:
merged_dfs = merged_dfs.drop(['number_of_open_complaints'], axis = 1)
merged_dfs

customer gender       state customer_lifetime_value   income  \
0     RB50392    NaN  Washington                     NaN      0.0   
1     QZ44356      F     Arizona              697953.59%      0.0   
2     AI49188      F      Nevada             1288743.17%  48767.0   
3     WW63253      M  California              764586.18%      0.0   
4     GA49547      M  Washington              536307.65%  36357.0   
...       ...    ...         ...                     ...      ...   
7065  LA72316      M  California             23405.98798  71941.0   
7066  PK87824      F  California             3096.511217  21604.0   
7067  TD14365      M  California             8163.890428      0.0   
7068  UP19263      M  California             7524.442436  21941.0   
7069  Y167826      M  California             2611.836866      0.0   

      monthly_premium_auto     policy_type  vehicle_class  total_claim_amount  
0                   1000.0   Personal Auto  Four-Door Car            2.704934  
1                     94.0   Personal Auto  Four-Door Car         1131.464935  
2                    108.0   Personal Auto   Two-Door Car          566.472247  
3                    106.0  Corporate Auto            SUV          529.881344  
4                     68.0   Personal Auto  Four-Door Car           17.269323  
...                    ...             ...            ...                 ...  
7065                  73.0   Personal Auto  Four-Door Car          198.234764  
7066                  79.0  Corporate Auto  Four-Door Car          379.200000  
7067                  85.0  Corporate Auto  Four-Door Car          790.784983  
7068                  96.0   Personal Auto  Four-Door Car          691.200000  
7069                  77.0  Corporate Auto   Two-Door Car          369.600000  

[12074 rows x 9 columns]

In [73]:
# Correct the values in the column customer lifetime value. They are given as a percent, so multiply them by 100 and change `dtype` to `numerical` type.
df['customer_lifetime_value'] = pd.to_numeric(df['customer_lifetime_value'].str.rstrip('%')) * 100
df['customer_lifetime_value']

0               NaN
1        69795359.0
2       128874317.0
3        76458618.0
4        53630765.0
           ...     
7065            NaN
7066            NaN
7067            NaN
7068            NaN
7069            NaN
Name: customer_lifetime_value, Length: 12074, dtype: float64

In [68]:
# Check for duplicate rows in the data and remove if any.
merged_dfs.duplicated()
merged_dfs.drop_duplicates()

customer gender       state customer_lifetime_value   income  \
0     RB50392    NaN  Washington                     NaN      0.0   
1     QZ44356      F     Arizona              697953.59%      0.0   
2     AI49188      F      Nevada             1288743.17%  48767.0   
3     WW63253      M  California              764586.18%      0.0   
4     GA49547      M  Washington              536307.65%  36357.0   
...       ...    ...         ...                     ...      ...   
7065  LA72316      M  California             23405.98798  71941.0   
7066  PK87824      F  California             3096.511217  21604.0   
7067  TD14365      M  California             8163.890428      0.0   
7068  UP19263      M  California             7524.442436  21941.0   
7069  Y167826      M  California             2611.836866      0.0   

      monthly_premium_auto     policy_type  vehicle_class  total_claim_amount  
0                   1000.0   Personal Auto  Four-Door Car            2.704934  
1                     94.0   Personal Auto  Four-Door Car         1131.464935  
2                    108.0   Personal Auto   Two-Door Car          566.472247  
3                    106.0  Corporate Auto            SUV          529.881344  
4                     68.0   Personal Auto  Four-Door Car           17.269323  
...                    ...             ...            ...                 ...  
7065                  73.0   Personal Auto  Four-Door Car          198.234764  
7066                  79.0  Corporate Auto  Four-Door Car          379.200000  
7067                  85.0  Corporate Auto  Four-Door Car          790.784983  
7068                  96.0   Personal Auto  Four-Door Car          691.200000  
7069                  77.0  Corporate Auto   Two-Door Car          369.600000  

[9135 rows x 9 columns]

In [72]:
# Filter out the data for customers who have an income of 0 or less.
df = merged_dfs

filtered_df = df[['customer', 'income']][df['income'] <= 0]
filtered_df 

customer  income
0     RB50392     0.0
1     QZ44356     0.0
3     WW63253     0.0
7     CF85061     0.0
10    SX51350     0.0
...       ...     ...
7059  WZ45103     0.0
7061  RX91025     0.0
7062  AC13887     0.0
7067  TD14365     0.0
7069  Y167826     0.0

[2294 rows x 2 columns]